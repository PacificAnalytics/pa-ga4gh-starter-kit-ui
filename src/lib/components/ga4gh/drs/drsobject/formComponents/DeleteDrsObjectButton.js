import React, {useState} from 'react';
import { 
    Typography,
    Button, 
    Dialog, 
    DialogTitle, 
    DialogContent, 
    DialogContentText, 
    DialogActions
} from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';
import {
    useHistory
} from 'react-router-dom';
import ApiCaller from '../../../../../utils/ApiCaller';
import { scrollToTop } from '../../../../../functions/common';

const DeleteDrsObjectButton = props => {
    let history = useHistory();
    const [confirmationDialogIsOpen, setConfirmationDialogIsOpen] = useState(false);
    const closeConfirmationDialog = () => {
        setConfirmationDialogIsOpen(false);
    }
    const deleteDrsObject = (id, setSuccessMessage, setError, retrieveDrsObjectsList) => {
        let requestConfig = {
                url: `http://localhost:8080/admin/ga4gh/drs/v1/objects/${id}`,
                method: 'DELETE'
        }
        ApiCaller(
            requestConfig, 
            responseData => {
                setSuccessMessage(`DRS Object '${id}' has been successfully deleted`);
                retrieveDrsObjectsList();
                history.push('/drs');
                scrollToTop();
            }, 
            error => {
                setError(error);
            }
        )
        setConfirmationDialogIsOpen(false);
    }

    return (
        <div>
            <Button variant='contained' color='secondary' size='large' aria-label='delete-drs-object-button'
                endIcon={<DeleteIcon/>} onClick={() => setConfirmationDialogIsOpen(true)}>
                <Typography variant='button'>Delete</Typography>
            </Button> 
            <Dialog open={confirmationDialogIsOpen} onClose={closeConfirmationDialog}>
                <DialogTitle align='center'>Are you sure you want to delete the DRS Object?</DialogTitle>
                <DialogContent>
                    <DialogContentText align='center'>
                        Deleting this DRS Object will permanently remove it from the database.
                    </DialogContentText>
                </DialogContent>
                <DialogActions>
                    <Button variant='contained' color='primary' onClick={closeConfirmationDialog}>
                        <Typography variant='button'>Cancel</Typography>
                    </Button>
                    <Button variant='contained' color='primary' 
                    onClick={event => deleteDrsObject(props.id, props.setSuccessMessage, props.setError, props.retrieveDrsObjectsList)}>
                        <Typography variant='button'>Delete</Typography>
                    </Button>
                </DialogActions>
            </Dialog>
        </div>
    )
}

export default DeleteDrsObjectButton;