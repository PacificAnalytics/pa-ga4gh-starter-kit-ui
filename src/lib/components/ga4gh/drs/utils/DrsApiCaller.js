import axios from 'axios';

const DrsApiCaller = (requestConfig, handleResponseData, handleError) => {
  axios(requestConfig)
  .then(response => {
    handleResponseData(response.data);
  })
  .catch(error => {
    handleError(error)
  })
}

export default DrsApiCaller;
