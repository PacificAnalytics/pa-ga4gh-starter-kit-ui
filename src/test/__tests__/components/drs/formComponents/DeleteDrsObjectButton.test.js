import React from 'react';
import renderer from 'react-test-renderer';
import {getByRole, getByText, render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import {mockValidTestBlob} from '../../../../resources/MockData';
import DeleteDrsObjectButton from '../../../../../lib/components/drs/formComponents/DeleteDrsObjectButton';import {
    setSuccessMessage, 
    retrieveDrsObjectsList, 
    setError
} from '../../../../resources/MockFunctions';

jest.mock('../../../../../lib/components/drs/utils/DrsApiCaller');
import DrsApiCaller from '../../../../../lib/components/drs/utils/DrsApiCaller';

let deleteMockFunctions = null;

beforeEach(() => {
    deleteMockFunctions = {
        setSuccessMessage: setSuccessMessage,
        retriveDrsObejctsList: retrieveDrsObjectsList,
        setError: setError,
    }
});

afterEach(() => {
    setSuccessMessage.mockClear();
    retrieveDrsObjectsList.mockClear();
    setError.mockClear();
})

test('<DeleteDrsObjectButton /> can delete DRS Object successfully', () => {
    DrsApiCaller.mockImplementation(() => {
        setSuccessMessage(`DRS Object '${mockValidTestBlob.id}' has been successfully deleted`);
        retrieveDrsObjectsList();
    });
    
    let {container} = render(<DeleteDrsObjectButton id={mockValidTestBlob.id} {...deleteMockFunctions}/> );
    expect(container.firstChild).toMatchSnapshot();
    let deleteButton = screen.getByRole('button');
    expect(deleteButton).toBeInTheDocument();
    userEvent.click(deleteButton);
    let confirmationDialog = screen.getByRole('dialog');
    expect(confirmationDialog).toBeVisible();
    expect(getByText(confirmationDialog, 'Are you sure you want to delete the DRS Object?')).toBeInTheDocument();
    expect(getByText(confirmationDialog, 'Deleting this DRS Object will permanently remove it from the database.')).toBeInTheDocument();
    expect(getByRole(confirmationDialog, 'button', {name: 'Cancel'})).toBeInTheDocument();
    expect(getByRole(confirmationDialog, 'button', {name: 'Delete'})).toBeInTheDocument();

    // clicking cancel button should close dialog
    userEvent.click(getByRole(confirmationDialog, 'button', {name: 'Cancel'}));
    expect(confirmationDialog).not.toBeVisible();

    // reopening dialog and clicking delete button should delete DRS Object
    userEvent.click(deleteButton);
    userEvent.click(getByRole(confirmationDialog, 'button', {name: 'Delete'}));
    expect(setSuccessMessage).toHaveBeenCalled();
    expect(retrieveDrsObjectsList).toHaveBeenCalled();
    expect(setError).not.toHaveBeenCalled();
});

test('<DeleteDrsObjectButton /> can handle errors', () => {
    DrsApiCaller.mockImplementation(() => {
        setError('Error: DRS object was not deleted successfully')
    });
    
    let {container} = render(<DeleteDrsObjectButton id={mockValidTestBlob.id} {...deleteMockFunctions}/> );
    let deleteButton = screen.getByRole('button');
    userEvent.click(deleteButton);
    let confirmationDialog = screen.getByRole('dialog');
    expect(confirmationDialog).toBeVisible();
    // clicking delete button should cause error
    userEvent.click(deleteButton);
    userEvent.click(getByRole(confirmationDialog, 'button', {name: 'Delete'}));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});