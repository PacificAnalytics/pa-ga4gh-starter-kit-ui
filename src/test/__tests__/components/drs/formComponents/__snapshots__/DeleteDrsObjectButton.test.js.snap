// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteDrsObjectButton /> can delete DRS Object successfully 1`] = `
<div>
  <button
    aria-label="delete-drs-object-button"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiTypography-root MuiTypography-button"
      >
        Delete
      </span>
      <span
        class="MuiButton-endIcon MuiButton-iconSizeLarge"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </span>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
